From ef46529db9e02a80d411b0a044857151eeecfb73 Mon Sep 17 00:00:00 2001
From: Pulkit077 <pulkitagarwal2k1@gmail.com>
Date: Sat, 21 Oct 2023 19:01:05 +0530
Subject: [PATCH 09/12] core: camera2: StreamConfigurationMap: add constructor
 for MIUI camera

This is a squash of the following commits.

 Author:    Pulkit077 <pulkitagarwal2k1@gmail.com>
 Date:      Sat Oct 21 19:01:05 2023 +0530

    core: camera2: StreamConfigurationMap: add constructor for MIUI camera

    neobuddy89: Updated for A15 QPR2

    Change-Id: Ie903c11161664153cb26858da426b546e655d3d2
    Signed-off-by: Pulkit077 <pulkitagarwal2k1@gmail.com>

 Author:    Pranav Vashi <neobuddy89@gmail.com>
 Date:      Fri Mar 28 00:12:49 2025 +0530

    camera2: StreamConfigurationMap: add another constructor for MIUI camera on A15 QPR2

    Log:
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat: Failed to construct config map.
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat: java.lang.NoSuchMethodException: android.hardware.camera2.params.StreamConfigurationMap.<init> [class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.HighSpeedVideoConfiguration;, class android.hardware.camera2.params.ReprocessFormatsMap, boolean]
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.lang.Class.getConstructor0(Class.java:3395)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.lang.Class.getDeclaredConstructor(Class.java:3077)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.o.i.b.c.b(SourceFile:5)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.b.g4.r1(SourceFile:21)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.b.g4.W1(SourceFile:4)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.b.g4.V1(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.b.h4.v1(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.l8.n1.C(SourceFile:2)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.l8.n1.p(SourceFile:8)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.gp(SourceFile:18)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.Nh(SourceFile:46)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.t6.b.i.c0.Nh(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.p7.Ha(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.lj(SourceFile:14)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.nj(SourceFile:6)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.t6.b.i.c0.nj(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.p7.Ma(SourceFile:2)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.o8.a.g.a(SourceFile:32)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.d.a.d7.o8.a.g.apply(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SourceFile:3)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.schedulers.ScheduledDirectTask.call(SourceFile:3)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.schedulers.ScheduledDirectTask.call(SourceFile:1)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.util.concurrent.FutureTask.run(FutureTask.java:317)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:348)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at d.o.f.u.k$b.run(SourceFile:4)
    03-26 18:07:26.280  8502  8705 W CameraConstructCompat:   at java.lang.Thread.run(Thread.java:1119)
    03-26 18:07:26.280  8502  8705 D CAM_CameraSizeManager: updateSize: The best sensor raw image size: 0x0
    03-26 18:07:26.280  8502  8705 D CAM_CameraCapabilities: addStreamConfigurationToList: xiaomi.scaler.availableLimitStreamConfigurations: size = 222
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat: Failed to construct config map.
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat: java.lang.NoSuchMethodException: android.hardware.camera2.params.StreamConfigurationMap.<init> [class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfiguration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.StreamConfigurationDuration;, class [Landroid.hardware.camera2.params.HighSpeedVideoConfiguration;, class android.hardware.camera2.params.ReprocessFormatsMap, boolean]
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.lang.Class.getConstructor0(Class.java:3395)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.lang.Class.getDeclaredConstructor(Class.java:3077)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.o.i.b.c.b(SourceFile:5)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.b.g4.r1(SourceFile:21)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.b.g4.W1(SourceFile:4)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.b.g4.V1(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.b.h4.v1(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.l8.n1.p(SourceFile:11)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.gp(SourceFile:18)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.Nh(SourceFile:46)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.t6.b.i.c0.Nh(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.p7.Ha(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.lj(SourceFile:14)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.q7.nj(SourceFile:6)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.t6.b.i.c0.nj(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.p7.Ma(SourceFile:2)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.o8.a.g.a(SourceFile:32)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.d.a.d7.o8.a.g.apply(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SourceFile:3)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.schedulers.ScheduledDirectTask.call(SourceFile:3)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at io.reactivex.internal.schedulers.ScheduledDirectTask.call(SourceFile:1)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.util.concurrent.FutureTask.run(FutureTask.java:317)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:348)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at d.o.f.u.k$b.run(SourceFile:4)
    03-26 18:07:26.281  8502  8705 W CameraConstructCompat:   at java.lang.Thread.run(Thread.java:1119)

    Change-Id: Ic80969b501a13a4b936aa192c07fcf57034c5fed
    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

Change-Id: I22f738960fcca9cdcb396744026e90af448f266d
Signed-off-by: pure-soul-kk <krishnakripa34567@gmail.com>
---
 .../params/StreamConfigurationMap.java        | 130 ++++++++++++++++++
 1 file changed, 130 insertions(+)

diff --git a/core/java/android/hardware/camera2/params/StreamConfigurationMap.java b/core/java/android/hardware/camera2/params/StreamConfigurationMap.java
index 323459d553ac..e7a8dd41da20 100644
--- a/core/java/android/hardware/camera2/params/StreamConfigurationMap.java
+++ b/core/java/android/hardware/camera2/params/StreamConfigurationMap.java
@@ -151,6 +151,136 @@ public final class StreamConfigurationMap {
                     listHighResolution, /*enforceImplementationDefined*/ true);
     }
 
+    /**
+     * Create a new {@link StreamConfigurationMap}.
+     *
+     * <p>The array parameters ownership is passed to this object after creation; do not
+     * write to them after this constructor is invoked.</p>
+     *
+     * @param configurations a non-{@code null} array of {@link StreamConfiguration}
+     * @param minFrameDurations a non-{@code null} array of {@link StreamConfigurationDuration}
+     * @param stallDurations a non-{@code null} array of {@link StreamConfigurationDuration}
+     * @param depthConfigurations a non-{@code null} array of depth {@link StreamConfiguration}
+     * @param depthMinFrameDurations a non-{@code null} array of depth
+     *        {@link StreamConfigurationDuration}
+     * @param depthStallDurations a non-{@code null} array of depth
+     *        {@link StreamConfigurationDuration}
+     * @param dynamicDepthConfigurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfiguration}
+     * @param dynamicDepthMinFrameDurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfigurationDuration}
+     * @param dynamicDepthStallDurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfigurationDuration}
+     * @param heicConfigurations a non-{@code null} array of heic {@link StreamConfiguration}
+     * @param heicMinFrameDurations a non-{@code null} array of heic
+     *        {@link StreamConfigurationDuration}
+     * @param heicStallDurations a non-{@code null} array of heic
+     *        {@link StreamConfigurationDuration}
+     * @param highSpeedVideoConfigurations an array of {@link HighSpeedVideoConfiguration}, null if
+     *        camera device does not support high speed video recording
+     * @param listHighResolution a flag indicating whether the device supports BURST_CAPTURE
+     *        and thus needs a separate list of slow high-resolution output sizes
+     * @throws NullPointerException if any of the arguments except highSpeedVideoConfigurations
+     *         were {@code null} or any subelements were {@code null}
+     *
+     * @hide
+     */
+    public StreamConfigurationMap(
+            StreamConfiguration[] configurations,
+            StreamConfigurationDuration[] minFrameDurations,
+            StreamConfigurationDuration[] stallDurations,
+            StreamConfiguration[] depthConfigurations,
+            StreamConfigurationDuration[] depthMinFrameDurations,
+            StreamConfigurationDuration[] depthStallDurations,
+            StreamConfiguration[] dynamicDepthConfigurations,
+            StreamConfigurationDuration[] dynamicDepthMinFrameDurations,
+            StreamConfigurationDuration[] dynamicDepthStallDurations,
+            StreamConfiguration[] heicConfigurations,
+            StreamConfigurationDuration[] heicMinFrameDurations,
+            StreamConfigurationDuration[] heicStallDurations,
+            HighSpeedVideoConfiguration[] highSpeedVideoConfigurations,
+            ReprocessFormatsMap inputOutputFormatsMap,
+            boolean listHighResolution) {
+        this(configurations, minFrameDurations, stallDurations,
+                    depthConfigurations, depthMinFrameDurations, depthStallDurations,
+                    dynamicDepthConfigurations, dynamicDepthMinFrameDurations,
+                    dynamicDepthStallDurations,
+                    heicConfigurations, heicMinFrameDurations, heicStallDurations,
+                    null /*jpegRConfigurations*/, null /*jpegRMinFrameDurations*/, null /*jpegRStallDurations*/,
+                    null /*heicUltraHDRConfigurations*/, null /*heicUltraHDRMinFrameDurations*/,
+                    null /*heicUltraHDRStallDurations*/, highSpeedVideoConfigurations, inputOutputFormatsMap,
+                    listHighResolution, /*enforceImplementationDefined*/ true);
+    }
+
+    /**
+     * Create a new {@link StreamConfigurationMap}.
+     *
+     * <p>The array parameters ownership is passed to this object after creation; do not
+     * write to them after this constructor is invoked.</p>
+     *
+     * @param configurations a non-{@code null} array of {@link StreamConfiguration}
+     * @param minFrameDurations a non-{@code null} array of {@link StreamConfigurationDuration}
+     * @param stallDurations a non-{@code null} array of {@link StreamConfigurationDuration}
+     * @param depthConfigurations a non-{@code null} array of depth {@link StreamConfiguration}
+     * @param depthMinFrameDurations a non-{@code null} array of depth
+     *        {@link StreamConfigurationDuration}
+     * @param depthStallDurations a non-{@code null} array of depth
+     *        {@link StreamConfigurationDuration}
+     * @param dynamicDepthConfigurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfiguration}
+     * @param dynamicDepthMinFrameDurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfigurationDuration}
+     * @param dynamicDepthStallDurations a non-{@code null} array of dynamic depth
+     *        {@link StreamConfigurationDuration}
+     * @param heicConfigurations a non-{@code null} array of heic {@link StreamConfiguration}
+     * @param heicMinFrameDurations a non-{@code null} array of heic
+     *        {@link StreamConfigurationDuration}
+     * @param heicStallDurations a non-{@code null} array of heic
+     *        {@link StreamConfigurationDuration}
+     * @param jpegRConfigurations a non-{@code null} array of Jpeg/R {@link StreamConfiguration}
+     * @param jpegRMinFrameDurations a non-{@code null} array of Jpeg/R
+     *        {@link StreamConfigurationDuration}
+     * @param jpegRStallDurations a non-{@code null} array of Jpeg/R
+     *        {@link StreamConfigurationDuration}
+     * @param highSpeedVideoConfigurations an array of {@link HighSpeedVideoConfiguration}, null if
+     *        camera device does not support high speed video recording
+     * @param listHighResolution a flag indicating whether the device supports BURST_CAPTURE
+     *        and thus needs a separate list of slow high-resolution output sizes
+     * @throws NullPointerException if any of the arguments except highSpeedVideoConfigurations
+     *         were {@code null} or any subelements were {@code null}
+     *
+     * @hide
+     */
+    public StreamConfigurationMap(
+            StreamConfiguration[] configurations,
+            StreamConfigurationDuration[] minFrameDurations,
+            StreamConfigurationDuration[] stallDurations,
+            StreamConfiguration[] depthConfigurations,
+            StreamConfigurationDuration[] depthMinFrameDurations,
+            StreamConfigurationDuration[] depthStallDurations,
+            StreamConfiguration[] dynamicDepthConfigurations,
+            StreamConfigurationDuration[] dynamicDepthMinFrameDurations,
+            StreamConfigurationDuration[] dynamicDepthStallDurations,
+            StreamConfiguration[] heicConfigurations,
+            StreamConfigurationDuration[] heicMinFrameDurations,
+            StreamConfigurationDuration[] heicStallDurations,
+            StreamConfiguration[] jpegRConfigurations,
+            StreamConfigurationDuration[] jpegRMinFrameDurations,
+            StreamConfigurationDuration[] jpegRStallDurations,
+            HighSpeedVideoConfiguration[] highSpeedVideoConfigurations,
+            ReprocessFormatsMap inputOutputFormatsMap,
+            boolean listHighResolution) {
+        this(configurations, minFrameDurations, stallDurations,
+                    depthConfigurations, depthMinFrameDurations, depthStallDurations,
+                    dynamicDepthConfigurations, dynamicDepthMinFrameDurations,
+                    dynamicDepthStallDurations,
+                    heicConfigurations, heicMinFrameDurations, heicStallDurations,
+                    jpegRConfigurations, jpegRMinFrameDurations, jpegRStallDurations,
+                    null /*heicUltraHDRConfigurations*/, null /*heicUltraHDRMinFrameDurations*/,
+                    null /*heicUltraHDRStallDurations*/, highSpeedVideoConfigurations, inputOutputFormatsMap,
+                    listHighResolution, /*enforceImplementationDefined*/ true);
+    }
+
     /**
      * Create a new {@link StreamConfigurationMap}.
      *
-- 
2.53.0

